### PR TITLE
Register REST API and forward REST requests to associated extension

### LINKED ISSUE #64: [FEATURE] [META] Register RestHandlers in OpenSearch
### Is your feature request related to a problem?
Currently, extensions are initialized in OpenSearch through extensions.yml file. For OpenSearch to process any functionality of the extension be it creating an index, deleting it or any other extension's functionality, the REST APIs of the extension should be registered in OpenSearch (ExtensionsOrchestrator) for OpenSearch to understand which extension it should send request to after receiving it from the user. The RestHandler should do the job of registering those APIs and map it in the sample format mentioned below
```
{extensionNode1: [API1, API2, API3], extenionNode2: [API1, API4, API5]}
```
extensionNode will be the node of the extension and values will be the APIs of the extension.

This issue requires us to
1. Register extension REST APIs to ExtensionOrchestrator by building Rest Handler model [3][4]
2. Map the extension node with the APIs it has  to register [6] [7]
3. Process the API call sent by the user and send a transport request to the right extension which has the request registered for that API [5] [8]

Rest_new

### What solution would you like?
_A clear and concise description of what you want to happen._

### What alternatives have you considered?
_A clear and concise description of any alternative solutions or features you've considered._

### Do you have any additional context?
_Add any other context or screenshots about the feature request here._

### LINKED ISSUE #69: [FEATURE] Register REST API and map to associated extension in ExtensionsOrchestrator
### Is your feature request related to a problem?
See steps 4, 6, and 7 described in #64.

### What solution would you like?
Upon receipt of REST API listing provided in #68, The `ExtensionsOrchestrator` should:
 * Register the APIs with the `RestHandler`
 * Store the response from the RestHandler in a map with the API as the key and the requesting extension as the value.

### What alternatives have you considered?
The best location for the map and API redirects to be stored is still an open question.  

### Do you have any additional context?
See also the DESIGN.md file (draft in #60)

### LINKED ISSUE #70: [FEATURE] Redirect REST API request to appropriate extensions
### Is your feature request related to a problem?
See steps 5, 8, and 9 in #64.

### What solution would you like?
Following registration, incoming API requests associated with extensions will go to the ExtensionsOrchestrator, where they will be sent to the appropriate extension to handle.  The extension will take the appropriate action for that API request, which may include one or more API requests to OpenSearch.

### What alternatives have you considered?
Data-heavy API requests likely do not need to be forwarded in their entirety. To minimize latency, it may be useful to separate an API request from its data in some situations.

### Do you have any additional context?
See also the DESIGN.md file in #60

### PR DESCRIPTION
**Pull request by dbwiddis**
Companion PR: https://github.com/opensearch-project/OpenSearch/pull/4282

### Description

This and the companion PR linked above completes the work begun in #74 by registering the received API with the `RestController`.  When OpenSearch receives a `RestRequest` matching one of these registered routes, it forwards it to the appropriate extension for further action.

See the design schematic and description in #64
 - Steps 1 and 2 were already completed
 - [#4100](https://github.com/opensearch-project/OpenSearch/pull/4100) and #74 implemented step 3
 - This and the companion PR implement steps 4 through 8
 - Step 9 is outside the scope of this PR.

### Issues Resolved
Closes #64, Closes #69, and Closes #70

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dbwiddis** [comment]
@saratvemulapalli @owaiskazi19 ready for review. Failing gradle check due to changes on companion PR; passes locally after publishToMavenLocal of the companion PR.

File: src/main/java/org/opensearch/sdk/ExtensionsRunner.java [hunk 1]
```diff
@@ -197,6 +199,24 @@ ExtensionBooleanResponse handleIndicesModuleNameRequest(IndicesModuleRequest ind
...
+     * @return A response acknowledging the request.
+     */
+    RestExecuteOnExtensionResponse handleRestExecuteOnExtensionRequest(RestExecuteOnExtensionRequest request) {
+        String message = "The extension would have just executed " + request.getMethod() + " " + request.getUri();
+        // TODO: logic matching the method/URI which came from the extension API to the appropriate
+        // action class, e.g., see AD plugin's actions in org.opensearch.ad.rest package.
+        // This should probably be stored locally in a map before we send the register API requests to OpenSearch.
+        // Tricky part is how to match up API names (in text) with actions (Class Names). Could use reflection
+        // or just register like the existing plugin code does. TBD future code!
+        // For now we just log (locally) and respond to OpenSearch that we received enough info to correlate such an action
```
Comment (owaiskazi19) [inline src/main/java/org/opensearch/sdk/ExtensionsRunner.java:215]: Can we open an issue for this TODO which will basically help the extension to redirect the request to the appropriate actions?

File: DESIGN.md [hunk 2]
```diff
@@ -51,16 +54,41 @@ extensions:
...
+
+### OpenSearch SDK for Java
+
+Currently, plugins rely on extension points to communicate with OpenSearch. These are represented as Actions. To turn plugins into extensions, the Extension must assemble a list of all methods and URIs to communicate to OpenSearch, where they will be registered; upon receiving a matching request from a user these will be forwarded back to the Extension and the Extension will further need to handle these registered methods and URIs with an appropriate Action.
+
+### Extension Walk Through
+
+1. Extensions are started up and must be running before OpenSearch is started.  (In the future, there will be a facility to refresh the extension list during operation and handle network communication interruptions.)
+
+2. OpenSearch is started. During its bootstrap, the `ExtensionsOrchestrator` is initialized, reading extension definitions from `extensions.yml`.
```
Comment (owaiskazi19) [inline DESIGN.md:75]:
```suggestion
2. OpenSearch is started. During its bootstrap, the `ExtensionsOrchestrator` is initialized, reading list of extension present in `extensions.yml`.
```

File: src/main/java/org/opensearch/sdk/ExtensionsRunner.java (hunk 1 above)
Comment (dbwiddis) [reply src/main/java/org/opensearch/sdk/ExtensionsRunner.java:215]: #102

**Review by saratvemulapalli — APPROVED** [review]
Thanks @dbwiddis for the changes!

**Review by owaiskazi19 — APPROVED** [review]
LGTM! Thanks for accommodating changes in DESIGN.md

**Review by joshpalis — APPROVED** [review]

**Pull request merged**